Prevent UserCleanupScheduler from overwhelming streaming

ThisIsMissEm opened the pull request:
UserCleanupScheduler is only processing statuses that have been deleted more than 30 days ago, as such, the likelihood of these statuses still being in any streams is quite low, so this change ensures that we don't have it publish a heap of useless messages to redis.

A single run of UserCleanupScheduler could previously have been amplifying traffic throughout the infrastructure of redis/sidekiq/streaming, and also causing excessive load on clients connected to streaming, as they'd be receiving a heap of messages about deleted statuses, even though those statuses had already been deleted a month earlier.

I did look into tests for this, however, there is no existing test coverage of the `redis.publish` calls from what I can see.

ThisIsMissEm committed:
Prevent UserCleanupScheduler from overwhelming streaming

ThisIsMissEm committed:
F^ Also deliver quietly any deletions of reblogs

ClearlyClaire commented on app/services/remove_status_service.rb:
```diff
@@ -12,6 +12,7 @@ class RemoveStatusService < BaseService
   # @option  [Boolean] :immediate
   # @option  [Boolean] :preserve
   # @option  [Boolean] :original_removed
+  # @option  [Boolean] :deliver_quietly
```
Maybe call it something like `skip_redis`?

Also I think instead of having the check in every method, I'd have the method calls grouped under a `unless skip_redis?` condition.

ThisIsMissEm replied:
I went with doing it in each method as each method might perform additional logic we do want executing, for instance, `remove_from_hashtags` decrements some counters, and we still want that to happen.

I wasn't sure if something like `skip_redis` was clear enough? But it's easy enough to change if we want a different name for it.

ClearlyClaire replied:
> I went with doing it in each method as each method might perform additional logic we do want executing, for instance, remove_from_hashtags decrements some counters, and we still want that to happen.

You're right, I overlooked that.

ThisIsMissEm replied:
What about `delete_quietly` instead of `deliver_quietly` ? Entirely up to you, I personally don't think there is a good name for this option.

ClearlyClaire replied:
`DeleteAccountService` has `skip_side_effects` and `skip_activitypub`, so I think `skip_redis` makes sense here, if a bit low-level.

ThisIsMissEm replied:
Maybe `skip_realtime` or `skip_streaming` would make most sense whilst not being `skip_redis` specifically.

i think I like `skip_streaming` the best.

ClearlyClaire replied:
I'm fine with either `skip_streaming` or `skip_redis`.

ClearlyClaire reviewed (commented):
> UserCleanupScheduler is only processing statuses that have been deleted more than 30 days ago, as such, the likelihood of these statuses still being in any streams is quite low, so this change ensures that we don't have it publish a heap of useless messages to redis.

Furthermore: the posts have *already* been soft-deleted, so it might be reasonable to not even call `RemoveStatusService` at all but destroy them all in one go.

That being said, I'm more comfortable with possibly doing some duplicate work when deleting posts than missing a code path and failing to delete a post, so this PR is fine by me.

ClearlyClaire commented on app/services/remove_status_service.rb:
```diff
@@ -103,6 +109,8 @@ def remove_reblogs
     # without us being able to do all the fancy stuff
 
     @status.reblogs.rewhere(deleted_at: [nil, @status.deleted_at]).includes(:account).reorder(nil).find_each do |reblog|
+      # We don't pass through deliver_quietly? here as the Reblog may be much
+      # newer than the status being removed
       RemoveStatusService.new.call(reblog, original_removed: true)
```
I think you should pass it. While reblogs may be much newer, it's supposed to have been deleted 30 days ago, like the original post being deleted.

ThisIsMissEm replied:
Apparently I didn't push the commit that resolves this, will fix later today!

ThisIsMissEm committed:
F^ rename deliver_quietly to skip_streaming

ThisIsMissEm commented:
Have updated and renamed the option to `skip_streaming` and addressed the issue with not passing it through for reblogs (it now does pass on the skip_streaming option to reblogs when removing them)

ThisIsMissEm commented:
@ClearlyClaire are you good with this being a squash merge, or should I rebase into a single commit?

ClearlyClaire commented:
We systematically squash merge.
